src/ripple/basics/scope.h

```diff
@@ -0,0 +1,196 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2021 Ripple Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_BASICS_SCOPE_H_INCLUDED
+#define RIPPLE_BASICS_SCOPE_H_INCLUDED
+
+#include <exception>
+#include <type_traits>
+#include <utility>
+
+namespace ripple {
+
+// RAII scope helpers.  As specified in Library Fundamental, Version 3
+// Basic design of idea:  https://www.youtube.com/watch?v=WjTrfoiB0MQ
+// Specification:
+//   http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/n4873.html#scopeguard
+
+template <class EF>
+class scope_exit
+{
+    EF exit_function_;
+    bool execute_on_destruction_{true};
+
+public:
+    ~scope_exit()
+    {
+        if (execute_on_destruction_)
+            exit_function_();
+    }
+
+    scope_exit(scope_exit&& rhs) noexcept(
+        std::is_nothrow_move_constructible_v<EF> ||
+        std::is_nothrow_copy_constructible_v<EF>)
+        : exit_function_{std::forward<EF>(rhs.exit_function_)}
```

Comment:
Since `rhs` is an rvalue, not a universal ref, shouldn't this be `std::move`, not `std::forward`?

Reply:
`forward` is used because `EF` is allowed to be an lvalue reference to a functor.  So this is a move if `EF` is an object, and a reference binding if `EF` is an lvalue reference.  With `std::move` this would fail to compile if `EF` is a lvalue reference.

Reply:
Makes sense. Thanks for the explanation.